Add metadata attribute to QuantumCircuit

mtreinish opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This commit adds a new attribute, metadata, to the QuantumCircuit class.
This attribute is used for storing arbitrary key value metadata about
the circuit as a dict. The intent is that this will be used by
downstream programs to keep track of multiple circuits as part of a
larger experiment. This metadata will not be used for anything related
to the operation of the circuit, it is just a free form field for users
to track circuits with richer metadata than just a name.

### Details and comments

Fixes #5237

chriseclectic commented on qiskit/assembler/assemble_circuits.py:
```diff
@@ -69,14 +69,18 @@ def _assemble_circuit(
                                   name=circuit.name,
-                                  global_phase=float(circuit.global_phase))
+                                  global_phase=float(circuit.global_phase),
+                                  **metadata)
```
It might be better to leave this as `metadata=metadata` so that we can pull out which values were added by the circuit metadata later (and not have memory slots etc get added into it). Also some metadata could potentially clash with other parameters in here.

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit-terra/pull/5270/commits/26036dc51ee57c5d4a7dd6bd866d73bf07233cb7

chriseclectic commented on qiskit/circuit/quantumcircuit.py:
```diff
@@ -137,7 +141,7 @@ class QuantumCircuit:
     extension_lib = "include \"qelib1.inc\";"
 
-    def __init__(self, *regs, name=None, global_phase=0):
+    def __init__(self, *regs, name=None, global_phase=0, **metadata):
```
Do we want to add it during construction, or only assign it as an attribute later? If we keep it here maybe it would be good to keep this an explicity kwarg `metadata=None` rather than arbitrary kwargs.

1ucian0 replied:
+1 to `metadata=None`

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit-terra/pull/5270/commits/26036dc51ee57c5d4a7dd6bd866d73bf07233cb7

chriseclectic commented on qiskit/circuit/quantumcircuit.py:
```diff
@@ -239,6 +244,24 @@ def data(self, data_input):
+        return self._metadata
+
+    @metadata.setter
+    def metadata(self, metadata):
```
Maybe we should do a type check here that metadata is a dict

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit-terra/pull/5270/commits/26036dc51ee57c5d4a7dd6bd866d73bf07233cb7

chriseclectic reviewed (changes requested):
One thing missing here is how to access this metadata through the result object as there is no way in the API to get the header information (iirc). You currently have to do `result.results[i].header` (which will also be populated with other added info like clbit_labels, qreg_sizes etc).

I think it would be best to add it to the qobj experiment header under a specific `metadata` or `circuit_metdata` key so you could have a method like:
`result.metadata(i) -> getattr(result.results[i].header, 'metadata', {})`

1ucian0 commented:
The way that `circuit.name` is carried over does not involve keeping it from pass to pass. The passmanager (qiskit/transpiler/runningpassmanager.py:104) adds the property based on the input:
```python
        name = circuit.name     # <---- keep the input circuit name
        dag = circuit_to_dag(circuit)
        ....
        for passset in self.working_list:
            for pass_ in passset:
                dag = self._do_pass(pass_, dag, passset.options)
        ...
        circuit = dag_to_circuit(dag)
        ...
        circuit.name = name    # <---- set the name into the output circuit
        return circuit
```

Assuming that metadata wont be used by any pass (in that case maybe should be somewhere else), a similar strategy can be used here.

nonhermitian commented:
This is a good idea, and a long time coming.  I think it is more than just expanding the name though.  One thing that comes to mind is the final qubit permutation after the swap mapper is done.  Currently the only way to infer this permutation is to keep track of the measurements appended to a circuit.  

It would also be nice to have the option for metadata to be written to QASM output and restored when generating a circuit from a QASM file.

jyu00 commented:
Can we add `metadata` to `Schedule` as well? Today users can attach metadata to a circuit/schedule in a job via `QobjExperimentHeader`. But soon `backend.run()` will take circuits/schedules instead of Qobj, and users won't have a way to manipulate the headers. Having metadata in both circuit and schedule would be a good replacement. I don't know if anyone uses `QobjExperimentHeader` for pulse schedules today, but I also don't have evidence that no one does.  

Alternatively, `backend.run()` can take a list of "experiment_header", but the original issue #5237 points out the advantages of having the metadata in the circuit. And it'd be nice to have consistency for specifying metadata for both circuits and schedules.

nonhermitian commented:
Another thing to add here is the notion of context awareness to jobs / circuits.  That is to say that a job /circuits should know the runtime that sent them, e.g. VQE or QAOA from aqua.  With this knowledge those jobs and circuits can be batched and represented differently in the UI of IQX, and probably batch retrieved using the IBMQ provider.  I think @jyu00 already is working down this line of thought with her `JobManager`.

mtreinish commented:
Sorry for the delay I've updated this to factor in the review comments. I've updated the usage to make it a single dict kwarg and set as a metadata key in the qobj header. I also added it to the pulse schedule object so v2 providers can leverage metadata in place of what the qobj experiment header used to provide for both circuit and pulse jobs.

I decided to leave the transpiler logic in place as is because of @nonhermitian's comment of adding metadata to the dag in the transpiler for things like qubit permutation. I can actually see value in this as a  method to return arbitrary metadata about the transformations from a pass for pass authors.

taalexander commented on qiskit/pulse/schedule.py:
```diff
@@ -50,13 +50,20 @@ class Schedule(abc.ABC):
                                    Tuple[int, Union['Schedule', Instruction]]],
-                 name: Optional[str] = None):
+                 name: Optional[str] = None,
+                 metadata: Optional[dict] = None):
```
Thank you, this is great and can be used to transmit circuit header information to pulse assembly.

mtreinish replied:
Oh, that reminded me that I should have added the circuit metadata to the output of the pulse scheduler too. Right now it's not preserving it. I'll do that now

eggerdj commented on test/python/compiler/test_transpiler.py:
```diff
@@ -928,6 +928,23 @@ def test_no_infinite_loop(self, optimization_level):
+                [9, 8], [9, 10], [11, 3], [11, 10],
+                [11, 12], [12, 2], [13, 1], [13, 12]]
+
+        res = transpile(circuit, basis_gates=['id', 'p', 'sx', 'cx'],
+                        coupling_map=cmap,
+                        optimization_level=optimization_level)
```
Does this also test that `metadata` is properly passed around when converting to and from `dagdependency`?

mtreinish replied:
Err, no it it's not since the only pass using dagdependency is the template matching which isn't in the default optimization levels. I'll add a dedicated converter test verifying that it's preserved through conversion with dagdependency.

mtreinish replied:
Test coverage added in: https://github.com/Qiskit/qiskit-terra/pull/5270/commits/06140cd5e76de8372108127c792df8a2819a07a5

eggerdj reviewed (commented):
Looks fine to me.

chriseclectic commented on qiskit/circuit/quantumcircuit.py:
```diff
@@ -242,6 +249,26 @@ def data(self, data_input):
+        """Update the circuit metadata"""
+        if not isinstance(metadata, dict) and metadata is not None:
+            raise TypeError("Only a dictionary or None is accepted for circuit metadata")
+        self._metadata = metadata
```
Do we want to provide a way to incrementally add metadata to a circuit by inserting a key, val pair into the dict?

chriseclectic replied:
Nevermind, I just realized the property attribute doesn't prevent one from adding data via `circ.metadata[key] = val`

mtreinish replied:
You can do all of this already by operating on `circuit.metadata` directly. For example. `circuit.metadata['foo'] = 'bar'` will work fine because it's modifying the dictionary returned by the getter above. The setter is only used when you do a `circuit.metadata = foo` call to set the attribute to something

chriseclectic reviewed (approved):
LGTM, just one comment regarding possibly adding metadata incrementally.

One thing that looks to be missing is any documentation of how one would expect a backend to return metadata in the Result. Is it arbitrarily up to the backend to decide or do we want to build expectations into ExperimentResult object? Aer, for example, currently adds simulator produced metadata into experiment results but its difficult access since you have to do `Result_obj.results[k].metadata` to retrieve the dictionary.

mtreinish commented:
> LGTM, ~just one comment regarding possibly adding metadata incrementally.~
> 
> One thing that looks to be missing is any documentation of how one would expect a backend to return metadata in the Result. Is it arbitrarily up to the backend to decide or do we want to build expectations into ExperimentResult object? Aer, for example, currently adds simulator produced metadata into experiment results but its difficult access since you have to do `Result_obj.results[k].metadata` to retrieve the dictionary.

I'm thinking we can bake this into https://github.com/Qiskit/qiskit-terra/pull/5406 since we don't really have much documentation on provider expectations. I'll update either this or that based on which merges first to ensure we document how providers are expected to use metadata.
